File: ios/Classes/PipHandler.swift
import Foundation
import AVFoundation
import AVKit
import Flutter

/// Handles Picture-in-Picture functionality for iOS
@available(iOS 9.0, *)
class PipHandler: NSObject {
    private let playerId: String
    private let channel: FlutterMethodChannel
    private var player: AVPlayer?
    private var playerLayer: AVPlayerLayer?
    private var pipController: AVPictureInPictureController?
    private var isInPipMode: Bool = false
    
    // Configuration
    private var config: [String: Any]?
    
    init(playerId: String, channel: FlutterMethodChannel) {
        self.playerId = playerId
        self.channel = channel
        super.init()
    }
    
    // MARK: - Initialization
    
    func initialize(player: AVPlayer?, playerLayer: AVPlayerLayer?) {
        print("PipHandler: Initializing for player: \(playerId)")
        
        self.player = player
        self.playerLayer = playerLayer
        
        // Check if PiP is supported
        guard AVPictureInPictureController.isPictureInPictureSupported() else {
            print("PipHandler: Picture-in-Picture not supported on this device")
            notifyPipStatusChanged(
                state: "unavailable",
                isSupported: false,
                isActive: false,
                errorMessage: "PiP not supported on this device"
            )
            return
        }
        
        // Setup PiP controller if we have a player layer
        if let playerLayer = playerLayer {
            setupPipController(with: playerLayer)
        }
        
        print("PipHandler: Initialized successfully")
    }
    
    // MARK: - PiP Controller Setup
    
    private func setupPipController(with playerLayer: AVPlayerLayer) {
        guard AVPictureInPictureController.isPictureInPictureSupported() else {
            print("PipHandler: Cannot setup PiP controller - not supported")
            return
        }
        
        do {
            // Create PiP controller
            pipController = try AVPictureInPictureController(playerLayer: playerLayer)
            pipController?.delegate = self
            
            // Configure PiP controller
            if #available(iOS 14.2, *) {
                pipController?.canStartPictureInPictureAutomaticallyFromInline = config?["autoEnterOnBackground"] as? Bool ?? false
            }
            
            // Notify that PiP is available
            notifyPipStatusChanged(
                state: "available",
                isSupported: true,
                isActive: false
            )
            
            print("PipHandler: PiP controller created successfully")
        } catch {
            print("PipHandler: Failed to create PiP controller: \(error.localizedDescription)")
            notifyPipStatusChanged(
                state: "unavailable",
                isSupported: false,
                isActive: false,
                errorMessage: error.localizedDescription
            )
        }
    }
    
    // MARK: - PiP Availability
    
    func checkAvailability() -> Bool {
        let isSupported = AVPictureInPictureController.isPictureInPictureSupported()
        print("PipHandler: PiP availability check - Supported: \(isSupported)")
        
        notifyPipStatusChanged(
            state: isSupported ? "available" : "unavailable",
            isSupported: isSupported,
            isActive: isInPipMode
        )
        
        return isSupported
    }
    
    // MARK: - Enter/Exit PiP
    
    func enterPip(config: [String: Any]?) -> Bool {
        print("PipHandler: Attempting to enter PiP mode")
        
        guard AVPictureInPictureController.isPictureInPictureSupported() else {
            print("PipHandler: Cannot enter PiP - not supported")
            return false
        }
        
        self.config = config
        
        // Create PiP controller if not exists
        if pipController == nil {
            if let playerLayer = playerLayer {
                setupPipController(with: playerLayer)
            } else {
                print("PipHandler: Cannot enter PiP - no player layer available")
                notifyPipStatusChanged(
                    state: "unavailable",
                    isSupported: true,
                    isActive: false,
                    errorMessage: "No player layer available"
                )
                return false
            }
        }
        
        guard let pipController = pipController else {
            print("PipHandler: Cannot enter PiP - controller not available")
            return false
        }
        
        // Check if PiP can be started
        guard pipController.isPictureInPicturePossible else {
            print("PipHandler: Cannot enter PiP - not possible at this time")
            notifyPipStatusChanged(
                state: "failed",
                isSupported: true,
                isActive: false,
                errorMessage: "PiP not possible at this time"
            )
            return false
        }
        
        // Start PiP
        pipController.startPictureInPicture()
        print("PipHandler: PiP start requested")
        
        return true
    }
    
    func exitPip() {
        print("PipHandler: Exiting PiP mode")
        
        guard let pipController = pipController else {
            print("PipHandler: Cannot exit PiP - controller not available")
            return
        }
        
        if pipController.isPictureInPictureActive {
            pipController.stopPictureInPicture()
            print("PipHandler: PiP stop requested")
        } else {
            print("PipHandler: PiP is not active")
        }
    }
    
    // MARK: - Status
    
    func getStatus() -> [String: Any] {
        let isSupported = AVPictureInPictureController.isPictureInPictureSupported()
        
        return [
            "state": isInPipMode ? "active" : (isSupported ? "available" : "unavailable"),
            "isSupported": isSupported,
            "isActive": isInPipMode
        ]
    }
    
    func isActive() -> Bool {
        return isInPipMode
    }
    
    func isPossible() -> Bool {
        return pipController?.isPictureInPicturePossible ?? false
    }
    
    // MARK: - Flutter Communication
    
    private func notifyPipStatusChanged(
        state: String,
        isSupported: Bool,
        isActive: Bool,
        errorMessage: String? = nil
    ) {
        let statusMap: [String: Any?] = [
            "playerId": playerId,
            "state": state,
            "isSupported": isSupported,
            "isActive": isActive,
            "errorMessage": errorMessage
        ]
        
        DispatchQueue.main.async { [weak self] in
            self?.channel.invokeMethod("onPipStatusChanged", arguments: statusMap)
        }
    }
    
    // MARK: - Cleanup
    
    func dispose() {
        print("PipHandler: Disposing")
        
        // Stop PiP if active
        if isInPipMode {
            exitPip()
        }
        
        pipController?.delegate = nil
        pipController = nil
        playerLayer = nil
        player = nil
        config = nil
    }
    
    deinit {
        dispose()
    }
}

// MARK: - AVPictureInPictureControllerDelegate

@available(iOS 9.0, *)
extension PipHandler: AVPictureInPictureControllerDelegate {
    
    func pictureInPictureControllerWillStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("PipHandler: Will start Picture-in-Picture")
        
        notifyPipStatusChanged(
            state: "active",
            isSupported: true,
            isActive: false
        )
    }
    
    func pictureInPictureControllerDidStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("PipHandler: Did start Picture-in-Picture")
        
        isInPipMode = true
        
        notifyPipStatusChanged(
            state: "active",
            isSupported: true,
            isActive: true
        )
    }
    
    func pictureInPictureController(
        _ pictureInPictureController: AVPictureInPictureController,
        failedToStartPictureInPictureWithError error: Error
    ) {
        print("PipHandler: Failed to start Picture-in-Picture: \(error.localizedDescription)")
        
        isInPipMode = false
        
        notifyPipStatusChanged(
            state: "failed",
            isSupported: true,
            isActive: false,
            errorMessage: error.localizedDescription
        )
    }
    
    func pictureInPictureControllerWillStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("PipHandler: Will stop Picture-in-Picture")
        
        notifyPipStatusChanged(
            state: "exiting",
            isSupported: true,
            isActive: true
        )
    }
    
    func pictureInPictureControllerDidStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
        print("PipHandler: Did stop Picture-in-Picture")
        
        isInPipMode = false
        
        notifyPipStatusChanged(
            state: "available",
            isSupported: true,
            isActive: false
        )
    }
    
    func picture(_ pictureInPictureController: AVPictureInPictureController,
                restoreUserInterfaceForPictureInPictureStopWithCompletionHandler completionHandler: @escaping (Bool) -> Void) {
        print("PipHandler: Restore user interface for Picture-in-Picture stop")
        
        // Restore the player UI when returning from PiP
        // The app should restore its UI to show the video player
        
        // Call completion handler to indicate success
        completionHandler(true)
    }
    
    // iOS 14.0+
    @available(iOS 14.0, *)
    func pictureInPictureController(
        _ pictureInPictureController: AVPictureInPictureController,
        restoreUserInterfaceForPictureInPictureStop completionHandler: @escaping (Bool) -> Void
    ) {
        print("PipHandler: Restore user interface for Picture-in-Picture stop (iOS 14+)")
        
        // Restore the player UI when returning from PiP
        completionHandler(true)
    }
}

// MARK: - PiP Configuration Extension

extension PipHandler {
    /// Update PiP configuration
    func updateConfig(_ config: [String: Any]) {
        self.config = config
        
        // Update auto-start configuration if available
        if #available(iOS 14.2, *) {
            let autoStart = config["autoEnterOnBackground"] as? Bool ?? false
            pipController?.canStartPictureInPictureAutomaticallyFromInline = autoStart
            print("PipHandler: Updated auto-start configuration: \(autoStart)")
        }
    }
    
    /// Set whether PiP can start automatically
    @available(iOS 14.2, *)
    func setAutoStartEnabled(_ enabled: Bool) {
        pipController?.canStartPictureInPictureAutomaticallyFromInline = enabled
        print("PipHandler: Auto-start enabled: \(enabled)")
    }
    
    /// Check if PiP controller requires linear playback
    @available(iOS 14.0, *)
    func requiresLinearPlayback() -> Bool {
        return pipController?.requiresLinearPlayback ?? false
    }
}

File: android/src/main/kotlin/com/zionmedianetwork/zmedia_player/PipHandler.kt
package com.zionmedianetwork.zmedia_player

import android.app.Activity
import android.app.PendingIntent
import android.app.PictureInPictureParams
import android.app.RemoteAction
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.drawable.Icon
import android.os.Build
import android.util.Rational
import androidx.annotation.RequiresApi
import io.flutter.plugin.common.MethodChannel
import java.lang.ref.WeakReference

/**
 * Handles Picture-in-Picture mode for Android O and above
 */
class PipHandler(
    activity: Activity?,
    private val playerId: String,
    private val methodChannel: MethodChannel
) {
    companion object {
        private const val TAG = "PipHandler"

        // Explicit-component broadcast action for PipActionReceiver. Since
        // every PendingIntent we hand to the system targets the receiver by
        // component (setClass), this string is never matched via an
        // intent-filter -- it exists purely for logcat/debugging clarity.
        private const val ACTION_PIP_ACTION = "com.zionmedianetwork.zmedia_player.PIP_ACTION"

        // The system enforces a maximum number of visible PiP actions (3 on
        // API 26-31, 5 on API 32+ / 12L); actions beyond the max are simply
        // not shown by the system (no crash). We defensively cap at the
        // lowest common value so behaviour is consistent across every
        // supported API level rather than silently varying by OS version.
        private const val MAX_PIP_ACTIONS = 3

        // Registry of live PipHandler instances keyed by playerId.
        //
        // PipActionReceiver (below) is a manifest-registered BroadcastReceiver:
        // the OS instantiates it fresh for every broadcast, so it cannot hold a
        // reference to "its" PipHandler directly. Instead, each handler
        // publishes itself here on construction and removes itself in
        // dispose(), and the receiver looks the handler up by the "playerId"
        // extra carried on the PendingIntent built in buildActionPendingIntent().
        // Held as a WeakReference so a handler that failed to dispose() cleanly
        // for any reason cannot be kept alive by this registry (same pattern as
        // NotificationHandler.liveHandlers).
        private val liveHandlers = mutableMapOf<String, WeakReference<PipHandler>>()

        internal fun handlerFor(playerId: String): PipHandler? = liveHandlers[playerId]?.get()
    }

    init {
        liveHandlers[playerId] = WeakReference(this)
    }

    // Mutable so the plugin can refresh it on activity lifecycle callbacks
    // (rotation/config change). PipHandler instances are cached in a long-lived
    // map keyed by playerId, so the Activity captured at construction time can
    // otherwise become stale/destroyed after a configuration change while the
    // handler itself lives on. See updateActivity().
    private var activity: Activity? = activity

    private var config: Map<String, Any>? = null
    private var isInPipMode = false

    /**
     * Refresh the Activity reference held by this handler.
     *
     * Must be called by the owner (ZMediaPlayerPlugin) whenever its own
     * activity reference changes - i.e. from [ActivityAware.onAttachedToActivity]
     * and [ActivityAware.onReattachedToActivityForConfigChanges], and with `null`
     * from [ActivityAware.onDetachedFromActivity] - so that a cached PipHandler
     * never operates against a destroyed pre-rotation Activity.
     */
    fun updateActivity(activity: Activity?) {
        this.activity = activity
    }

    /**
     * Persist the PiP config map so that subsequent [enterPip] calls (and
     * [buildPipParams]) can read autoEnterOnBackground even if the caller
     * does not pass the config again at that point.
     */
    fun applyConfig(config: Map<String, Any>?) {
        if (config != null) {
            this.config = config
        }
    }

    /**
     * Check if PiP is available on this device
     */
    fun checkAvailability(): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            android.util.Log.d(TAG, "PiP not available: Android version ${Build.VERSION.SDK_INT} < O (26)")
            notifyPipStatusChanged(
                state = "unavailable",
                isSupported = false,
                isActive = false,
                errorMessage = "PiP requires Android 8.0 (API 26) or higher"
            )
            return false
        }

        val currentActivity = activity
        if (currentActivity == null) {
            android.util.Log.d(TAG, "PiP not available: No activity context")
            notifyPipStatusChanged(
                state = "unavailable",
                isSupported = false,
                isActive = false,
                errorMessage = "No activity available"
            )
            return false
        }

        val hasPipFeature = currentActivity.packageManager.hasSystemFeature(
            PackageManager.FEATURE_PICTURE_IN_PICTURE
        )

        android.util.Log.d(TAG, "PiP availability check: SDK=${Build.VERSION.SDK_INT}, hasFeature=$hasPipFeature, activity=true")

        // Notify status
        notifyPipStatusChanged(
            state = if (hasPipFeature) "available" else "unavailable",
            isSupported = hasPipFeature,
            isActive = isInPipMode
        )

        return hasPipFeature
    }

    /**
     * Enter Picture-in-Picture mode
     */
    fun enterPip(config: Map<String, Any>?): Boolean {
        android.util.Log.d(TAG, "Attempting to enter PiP mode")

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            android.util.Log.w(TAG, "PiP not supported on Android version < O")
            notifyPipStatusChanged(
                state = "failed",
                isSupported = false,
                isActive = false,
                errorMessage = "PiP requires Android 8.0 or higher"
            )
            return false
        }

        val currentActivity = activity
        if (currentActivity == null) {
            android.util.Log.w(TAG, "Cannot enter PiP: No activity context")
            notifyPipStatusChanged(
                state = "failed",
                isSupported = true,
                isActive = false,
                errorMessage = "No activity available"
            )
            return false
        }

        // Merge: caller-supplied config takes precedence; fall back to previously
        // stored config (primed by applyConfig during checkPipAvailability).
        val effectiveConfig = config ?: this.config
        this.config = effectiveConfig

        return try {
            val params = buildPipParams(effectiveConfig, currentActivity)
            android.util.Log.d(TAG, "Built PiP params, entering PiP mode...")

            val result = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                currentActivity.enterPictureInPictureMode(params)
            } else {
                false
            }

            if (result) {
                isInPipMode = true
                notifyPipStatusChanged(
                    state = "active",
                    isSupported = true,
                    isActive = true
                )
                android.util.Log.d(TAG, "Entered PiP mode successfully")
            } else {
                android.util.Log.w(TAG, "Activity.enterPictureInPictureMode() returned false")
                notifyPipStatusChanged(
                    state = "failed",
                    isSupported = true,
                    isActive = false,
                    errorMessage = "Failed to enter PiP mode"
                )
            }

            result
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Error entering PiP mode: ${e.message}", e)
            notifyPipStatusChanged(
                state = "failed",
                isSupported = true,
                isActive = false,
                errorMessage = e.message
            )
            false
        }
    }

    /**
     * Exit Picture-in-Picture mode
     */
    fun exitPip() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
            return
        }

        if (activity == null || !isInPipMode) {
            return
        }

        // The activity will automatically exit PiP when we request it
        // We can't programmatically exit PiP, but we can notify the state change
        isInPipMode = false
        notifyPipStatusChanged(
            state = "available",
            isSupported = true,
            isActive = false
        )

        android.util.Log.d(TAG, "Exited PiP mode")
    }

    /**
     * Handle PiP mode changed (called from activity)
     */
    fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean) {
        android.util.Log.d(TAG, "PiP mode changed: $isInPictureInPictureMode")

        isInPipMode = isInPictureInPictureMode

        notifyPipStatusChanged(
            state = if (isInPictureInPictureMode) "active" else "available",
            isSupported = true,
            isActive = isInPictureInPictureMode
        )
    }

    /**
     * Build PiP parameters
     */
    @RequiresApi(Build.VERSION_CODES.O)
    private fun buildPipParams(config: Map<String, Any>?, context: Context): PictureInPictureParams {
        val builder = PictureInPictureParams.Builder()

        // Set aspect ratio
        val aspectRatio = config?.get("aspectRatio") as? Double ?: (16.0 / 9.0)
        val rational = Rational(
            (aspectRatio * 100).toInt(),
            100
        )
        builder.setAspectRatio(rational)

        // PipConfig.actions / PipConfig.showPlaybackControls (Wave C, gate item
        // "PipConfig.actions and PipConfig.showPlaybackControls are ignored
        // natively"). RemoteAction and PictureInPictureParams.setActions() are
        // both available from API 26 (O) -- the same level this whole function
        // is already guarded at -- so no extra @RequiresApi is needed here.
        //
        // showPlaybackControls == false suppresses the actions entirely (no
        // setActions() call at all, matching "do not add playback actions");
        // when true (the default) the configured actions list, if any, is
        // rendered in the PiP overlay.
        val showPlaybackControls = config?.get("showPlaybackControls") as? Boolean ?: true
        if (showPlaybackControls) {
            val actionsRaw = config?.get("actions") as? List<*>
            val remoteActions = buildRemoteActions(context, actionsRaw)
            if (remoteActions.isNotEmpty()) {
                builder.setActions(remoteActions)
            }
        }

        // For Android 12+, we can add additional features
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            // Set auto-enter enabled
            val autoEnter = config?.get("autoEnterOnBackground") as? Boolean ?: false
            builder.setAutoEnterEnabled(autoEnter)

            // Set seamless resize enabled for smoother transitions
            builder.setSeamlessResizeEnabled(true)
        }

        return builder.build()
    }

    /**
     * Converts [PipConfig.actions] (a `List<Map>` produced by `PipAction.toMap()`
     * on the Dart side -- `lib/src/models/pip_config.dart`, shape
     * `{"id": String, "title": String, "icon": String?}`) into
     * [RemoteAction]s for [PictureInPictureParams.Builder.setActions].
     *
     * Icon resolution mirrors `NotificationHandler.resolveSmallIcon`: `icon` is
     * looked up as a drawable resource name in the host app's own resources;
     * if absent, blank, or unresolved, a stock system icon is used instead so
     * an action is never silently dropped just because its icon could not be
     * resolved.
     *
     * Each action's [PendingIntent] targets [PipActionReceiver] (declared in
     * this module's `AndroidManifest.xml`), which invokes `onPipAction` on the
     * method channel with `{"playerId": ..., "actionId": ...}` when tapped --
     * mirroring the `createAction`/`NotificationActionReceiver` mechanism in
     * `NotificationHandler.kt`.
     */
    @RequiresApi(Build.VERSION_CODES.O)
    private fun buildRemoteActions(context: Context, actionsRaw: List<*>?): List<RemoteAction> {
        if (actionsRaw.isNullOrEmpty()) return emptyList()

        return actionsRaw.mapNotNull { raw ->
            val map = raw as? Map<*, *> ?: return@mapNotNull null
            val id = map["id"] as? String ?: return@mapNotNull null
            val title = map["title"] as? String ?: id
            val icon = resolveActionIcon(context, map["icon"] as? String)

            RemoteAction(
                icon,
                title,
                title,
                buildActionPendingIntent(context, id)
            )
        }.take(MAX_PIP_ACTIONS)
    }

    /**
     * Resolves a [PipAction.icon] drawable resource name to an [Icon], falling
     * back to a stock system icon when the name is absent/blank or does not
     * resolve to a drawable in the host app's resources.
     */
    private fun resolveActionIcon(context: Context, name: String?): Icon {
        if (!name.isNullOrBlank()) {
            val resId = context.resources.getIdentifier(name, "drawable", context.packageName)
            if (resId != 0) {
                return Icon.createWithResource(context, resId)
            }
            android.util.Log.w(
                TAG,
                "PipAction icon '$name' was not found in the host app's drawable resources; " +
                    "using a fallback icon."
            )
        }
        return Icon.createWithResource(context, android.R.drawable.ic_media_play)
    }

    /**
     * Builds the broadcast [PendingIntent] for a single PiP action, explicitly
     * targeting [PipActionReceiver] and carrying this instance's [playerId]
     * plus the tapped action's [actionId].
     */
    private fun buildActionPendingIntent(context: Context, actionId: String): PendingIntent {
        val intent = Intent(ACTION_PIP_ACTION).apply {
            setClass(context, PipActionReceiver::class.java)
            putExtra(PipActionReceiver.EXTRA_PLAYER_ID, playerId)
            putExtra(PipActionReceiver.EXTRA_ACTION_ID, actionId)
        }

        return PendingIntent.getBroadcast(
            context,
            // Distinct request code per (playerId, actionId) so concurrent
            // instances' actions don't collide and overwrite each other's
            // PendingIntents via FLAG_UPDATE_CURRENT.
            31 * playerId.hashCode() + actionId.hashCode(),
            intent,
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
            } else {
                PendingIntent.FLAG_UPDATE_CURRENT
            }
        )
    }

    /**
     * Called by [PipActionReceiver] when a custom PiP action is tapped.
     * Forwards the tap to Flutter as `onPipAction`.
     */
    internal fun onPipActionSelected(actionId: String) {
        android.util.Log.d(TAG, "PiP action selected: $actionId")
        methodChannel.invokeMethod(
            "onPipAction",
            mapOf("playerId" to playerId, "actionId" to actionId)
        )
    }

    /**
     * Notify Flutter of PiP status changes
     */
    private fun notifyPipStatusChanged(
        state: String,
        isSupported: Boolean,
        isActive: Boolean,
        errorMessage: String? = null
    ) {
        val statusMap = mapOf(
            "playerId" to playerId,
            "state" to state,
            "isSupported" to isSupported,
            "isActive" to isActive,
            "errorMessage" to errorMessage
        )

        activity?.runOnUiThread {
            methodChannel.invokeMethod("onPipStatusChanged", statusMap)
        }
    }

    /**
     * Get current PiP status
     */
    fun getStatus(): Map<String, Any> {
        return mapOf(
            "state" to if (isInPipMode) "active" else if (checkAvailability()) "available" else "unavailable",
            "isSupported" to checkAvailability(),
            "isActive" to isInPipMode
        )
    }

    /**
     * Dispose the PiP handler
     */
    fun dispose() {
        android.util.Log.d(TAG, "Disposing PiP handler")
        liveHandlers.remove(playerId)
        config = null
        isInPipMode = false
    }
}

/**
 * Manifest-registered receiver (see `android/src/main/AndroidManifest.xml`)
 * that routes taps on custom PiP actions (see
 * [PipHandler.buildActionPendingIntent]) back to Flutter.
 *
 * Every [PendingIntent] that reaches this receiver is built by
 * [PipHandler.buildActionPendingIntent] with an explicit component target
 * (this class) and always carries [EXTRA_PLAYER_ID] plus [EXTRA_ACTION_ID].
 * The system instantiates a fresh receiver instance per broadcast, so it
 * cannot hold player state itself -- it looks the owning [PipHandler] up via
 * [PipHandler.handlerFor] and forwards the tap via
 * [PipHandler.onPipActionSelected], which invokes `onPipAction` on the method
 * channel with `{"playerId": ..., "actionId": ...}`.
 *
 * No host-app manifest changes are required: this receiver is declared in
 * this module's own `AndroidManifest.xml` and is merged into every consuming
 * app's manifest automatically by the Android Gradle build's manifest merger.
 */
class PipActionReceiver : android.content.BroadcastReceiver() {
    override fun onReceive(context: Context, intent: Intent) {
        val playerId = intent.getStringExtra(EXTRA_PLAYER_ID) ?: return
        val actionId = intent.getStringExtra(EXTRA_ACTION_ID) ?: return
        val handler = PipHandler.handlerFor(playerId) ?: return
        handler.onPipActionSelected(actionId)
    }

    companion object {
        const val EXTRA_PLAYER_ID = "playerId"
        const val EXTRA_ACTION_ID = "actionId"
    }
}
